src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -476,6 +484,85 @@ symbol_factoryt::get_deferred_size(irep_idt symbol_name) const
   return optional_lookup(deferred_array_sizes, symbol_name);
 }
 
+bool symbol_factoryt::is_string_array(const exprt &expr) const
+{
+  if(expr.id() != ID_symbol)
+  {
+    return false;
+  }
+  auto const &symbol_expr = to_symbol_expr(expr);
+  return object_factory_params.should_be_treated_as_string(
+    get_symbol_base_name(symbol_expr));
+}
+
+void symbol_factoryt::gen_nondet_string_init(
```

Comment:
Given that there is a lot of similarity to `gen_nondet_array_init`, is it worth thinking about a little bit of refactoring? The only differences between array and strings are:
1) null termination for strings
2) printable chars
However, 2 is not a *hard* constraint - theres nothing that says a C string has to contain printable chars, so this is effectively an optional restriction when initialising (and should probably be made a user option). Likewise, I could imagine array initialisation where you might want to specify a range for the values (not asking for that right now though!). Would it there for be better to have a generic `gen_nondet_array_init_with_constraint` option that takes a lamda/function pointer to do the actually element initialisation? You could then handle the null terminator case either by re-initialising the last element after the `gen_nondet_array_init_with_constraint` has been called (bit of a hack though...) or have that as an additional lambda/function pointer argument for 'final_element_initialiser' or something? You could then still keep these two `gen_nondet_string_init`/`gen_nondet_array_init` functions as simple wrappers around the `gen_nondet_array_init_with_constraint` function.

Reply:
Sounds reasonable to me, although rather than `with_constraint` I'd call it `with_initializer`, and the passed function just gets array index and array size as parameters (so any nasty position dependent stuff can be decided within the function)

Reply:
Yep, passing index + size sounds like a good solution.